File: attack.m
% Watermark Attack

clear all;

image_name='outputs/watermarked.png';
output_name='outputs/attacked.png';

% Attack image
image_object=imread(image_name);

Sh=size(image_object,1);
Sw=size(image_object,2);

% Detector
detector = WatermarkDetector;

% Random object
random_object=image_object;

while detector.detect(random_object) == 1

    x = randi([1 Sh],1);
    y = randi([1 Sw],1);
    random_object(x, y)=128;

end

%imwrite(random_object,output_name);

% Tangent
for ii = 1:Sh
    for jj = 1:Sw

        temp_object = random_object;
        pixel_color = temp_object(ii, jj);
        tangent(ii, jj) = 0;

        for kk = 1:3

            temp_object(ii, jj) = (pixel_color + (64 * kk)) mod 256;

            if detector.detect(temp_object) == 1
                tangent(ii, jj) = kk;
                break;
            end

        end

    end
end

display(tangent);

File: attack.r
require("MMSec")
source('detect.r')

# Load the watermark
watermark <- as.matrix(read.table("outputs/watermark", header=F))

# Load the image
image <- load.png("outputs/watermarked.png")
image <- as.matrix(image) * 255

attacked <- image

# Size of image
dim <- dim(image)
heigth <- dim[1]
width <- dim[2]

# Set random pixels to grey value until images is not detected anymore
repeat {
	watermarked <- detect(attacked, watermark)

	if (! watermarked) {
		break
	}

	x = sample(1:heigth, 1, replace=T)
	y = sample(1:width, 1, replace=T)

	attacked[x,y] = 128
}

# Save images
save.png(greymap(attacked), "outputs/attacked.png", overwrite=TRUE)
